Examples/Image/Detection/FastRCNN/BrainScript/A2_RunWithPyModel_distributed.py

```diff
@@ -0,0 +1,220 @@
+    fc_out    = fc_layers(roi_out)
+
+    # z = Dense(rois[0], num_classes, map_rank=1)(fc_out)  # --> map_rank=1 is not yet supported
+    W = parameter(shape=(4096, n_classes), init=glorot_uniform())
+    b = parameter(shape=n_classes, init=0)
+    z = times(fc_out, W) + b
+
+    return z
+
+# Trains a Fast R-CNN model
+def train_fast_rcnn(debug_output=False, model_path=model_file, num_quantization_bits=1, distributed_after=1):
+    if debug_output:
+        print("Storing graphs and intermediate models to %s." % os.path.join(abs_path, "Output"))
+
+    # Create the minibatch source
+    minibatch_source = create_mb_source(image_height, image_width, num_channels,
+                                        num_classes, num_rois, base_path, "train")
+
+    # Input variables denoting features, rois and label data
+    image_input = C.input_variable((num_channels, image_height, image_width))
+    roi_input   = C.input_variable((num_rois, 4))
+    label_input = C.input_variable((num_rois, num_classes))
+
+    # define mapping from reader streams to network inputs
+    input_map = {
+        image_input: minibatch_source.streams.features,
+        roi_input: minibatch_source.streams.rois,
+        label_input: minibatch_source.streams.roiLabels
+    }
+
+    # Instantiate the Fast R-CNN prediction model and loss function
+    frcn_output = frcn_predictor(image_input, roi_input, num_classes, model_path)
+    ce = cross_entropy_with_softmax(frcn_output, label_input, axis=1)
+    pe = classification_error(frcn_output, label_input, axis=1)
+    if debug_output:
+        plot(frcn_output, os.path.join(abs_path, "Output", "graph_frcn.png"))
+
+    # Set learning parameters
+    l2_reg_weight = 0.0005
+    lr_per_sample = [0.00001] * 10 + [0.000001] * 5 + [0.0000001]
+    lr_schedule = learning_rate_schedule(lr_per_sample, unit=UnitType.sample)
+    mm_schedule = momentum_as_time_constant_schedule(momentum_time_constant)
+
+    # Instantiate the trainer object for distributed learner
+    local_learner = momentum_sgd(frcn_output.parameters, lr_schedule, mm_schedule, l2_regularization_weight=l2_reg_weight)
+    learner = distributed.data_parallel_distributed_learner(local_learner, num_quantization_bits=num_quantization_bits, 
+                                                            distributed_after=warm_up)
+    progress_printer = ProgressPrinter(tag='Training', num_epochs=max_epochs)
+    trainer = Trainer(frcn_output, (ce, pe), learner, progress_printer)
+
+    # Get minibatches of images and perform model training
+    print("Training Fast R-CNN model for %s epochs." % max_epochs)
+    log_number_of_parameters(frcn_output)
+    for epoch in range(max_epochs):       # loop over epochs
+        sample_count = 0
+        while sample_count < epoch_size:  # loop over minibatches in the epoch
+            data = minibatch_source.next_minibatch(min(mb_size, epoch_size-sample_count), input_map=input_map)
```

Comment:
This would feed the same data to all workers, which effectively just scales up learning rate. Please use partitions.

Reply:
Are there any reference as to distributed learning? I didn't find the detailed specification on it.